messaging/src/main/java/org/axonframework/eventhandling/deadletter/DeadLetteringEventHandlerInvoker.java

```diff
@@ -116,33 +126,75 @@ public void handle(@Nonnull EventMessage<?> message, @Nonnull Segment segment) t
+                () -> new GenericDeadLetter<>(sequenceIdentifier, message)
+        )) {
             if (logger.isInfoEnabled()) {
                 logger.info("Event with id [{}] is added to the dead-letter queue "
                                     + "since its queue id [{}] is already present.",
                             message.getIdentifier(), sequenceIdentifier);
             }
+            markPresentInDLQ(sequenceIdentifier, segment);
         } else {
-            if (logger.isTraceEnabled()) {
-                logger.trace("Event [{}] with queue id [{}] is not present in the dead-letter queue."
-                                     + "Handle operation is delegated to the wrapped EventHandlerInvoker.",
-                             message, sequenceIdentifier);
+            if (!skipIfPresentCheck) {
+                markNotPresentInDLQ(sequenceIdentifier, segment);
             }
-            try {
-                super.invokeHandlers(message);
-            } catch (Exception e) {
-                DeadLetter<EventMessage<?>> letter = new GenericDeadLetter<>(sequenceIdentifier, message, e);
-                EnqueueDecision<EventMessage<?>> decision = enqueuePolicy.decide(letter, e);
-                if (decision.shouldEnqueue()) {
-                    Throwable cause = decision.enqueueCause().orElse(null);
-                    queue.enqueue(sequenceIdentifier, decision.withDiagnostics(letter.withCause(cause)));
-                } else if (logger.isInfoEnabled()) {
-                    logger.info("The enqueue policy decided not to dead letter event [{}].", message.getIdentifier());
-                }
+            invokeHandlers(message, segment, sequenceIdentifier);
+        }
+    }
+
+    private void invokeHandlers(@Nonnull EventMessage<?> message, @Nonnull Segment segment, Object sequenceIdentifier) {
+        if (logger.isTraceEnabled()) {
+            logger.trace("Event [{}] with queue id [{}] is not present in the dead-letter queue."
+                                 + "Handle operation is delegated to the wrapped EventHandlerInvoker.",
+                         message, sequenceIdentifier);
+        }
+        try {
+            super.invokeHandlers(message);
+        } catch (Exception e) {
+            DeadLetter<EventMessage<?>> letter = new GenericDeadLetter<>(sequenceIdentifier, message, e);
+            EnqueueDecision<EventMessage<?>> decision = enqueuePolicy.decide(letter, e);
+            if (decision.shouldEnqueue()) {
+                Throwable cause = decision.enqueueCause().orElse(null);
+                markPresentInDLQ(sequenceIdentifier, segment);
+                queue.enqueue(sequenceIdentifier, decision.withDiagnostics(letter.withCause(cause)));
+            } else if (logger.isInfoEnabled()) {
+                logger.info("The enqueue policy decided not to dead letter event [{}].", message.getIdentifier());
+            }
+        }
+    }
+
+    private boolean skipIfPresentCheck(
```

Comment:
Whe can't a domain event with sequence number 0 exist in the dlq? If the sequence identifier doesn't use the aggregate identifier, it may very well exist. I would skip this check and just treat those events like any other.

Reply:
Good points. I removed the check.